### PR TITLE
fix(nns-recovery): choose DFINITY-owned node as one with highest certification share height

### PR DESCRIPTION
**Pull request by pierugo-dfinity**
The introduction of the end-to-end NNS recovery test (https://github.com/dfinity/ic/commit/64da0373b9b0bbfa79b99cb4b55018f6ef94d29c) is very flaky. The reason is that in the case where the DFINITY-owned node is lagging behind (which happens when it is part of the faulty nodes), it could happen that its artifact pool has a finalization height which is lower than the highest certification share height across the subnet. In that case, we ask the `ICReplay` step of the recovery to replay until the latter but it will only replay until the highest finalized block.

Indeed, the recovery tool assumes that the node we download the state from (including the consensus artifact pools) from contains all artifacts (except certifications and certification shares) up to the highest certification share height. If this is not the case, we could download those artifacts from the relevant node. This is something that we take note of (CON-1580) and will look into in the future by allowing for example to download the artifacts from a different node than the state.

Note: this is not specific to NNS recovery, it is also the case in application subnet recoveries. In app subnet system tests, we avoid this edge-case by downloading the state from the highest certification height node. This PR thus does the same thing to fix the flakiness short-term. Once CON-1580 is implemented, we will go back to selecting the DFINITY-owned node randomly.

Note 2: This edge-case never happened in production in app subnet recoveries because the DFINITY-owned node always had up-to-date artifacts.

### CODE REVIEW

File: rs/tests/nested/src/lib.rs [hunk 1]
```diff
@@ -499,35 +477,19 @@ pub fn nns_recovery_test(env: TestEnv) {
...
-        next_step: None,
-    };
-
-    let mut subnet_recovery_tool =
-        NNSRecoverySameNodes::new(logger.clone(), recovery_args, subnet_args);
-
-    // Break f+1 nodes by SSHing into them and breaking the replica binary.
+    // Define faulty and healthy nodes, pick a DFINITY-owned node (could be healthy or faulty)
+    let mut nns_nodes = nns_subnet.nodes().collect::<Vec<_>>();
+    nns_nodes.shuffle(&mut rand::thread_rng());
```
Comment (frankdavid) [inline rs/tests/nested/src/lib.rs]: What's your goal with making this random? Usually, it'd preferred to make tests deterministic (or at least not introduce extra source of indeterminism).

File: rs/tests/nested/src/lib.rs [hunk 2]
```diff
@@ -576,21 +538,51 @@ pub fn nns_recovery_test(env: TestEnv) {
...
-                    .certification_share_height
-            })
-            .max()
-            .unwrap()
-            .get(),
+    // Choose the DFINITY-owned node to be the one with the highest certification share height
+    let (dfinity_owned_node, highest_certification_share_height) = nns_subnet
+        .nodes()
+        .filter_map(|n| {
+            block_on(get_node_metrics(&logger, &n.get_ip_addr()))
```
Comment (frankdavid) [inline rs/tests/nested/src/lib.rs:544]: How fast is get_node_metrics? Would it make sense to run the futures in parallel?

File: rs/tests/nested/src/lib.rs (hunk 1 above)
Comment (pierugo-dfinity) [reply rs/tests/nested/src/lib.rs]: Thank you very much for the remark. It was there before this PR to randomly select a DFINITY-owned that is healthy or faulty. It is not needed anymore. https://github.com/dfinity/ic/pull/6554/commits/d259485f89333229a94d734e8d39576c7c79cd52

P.S.: Additional test scenarios are planned, in which the DFINITY-owned node will be either always healthy or always faulty.

File: rs/tests/nested/src/lib.rs (hunk 2 above)
Comment (pierugo-dfinity) [reply rs/tests/nested/src/lib.rs:544]: `get_node_metrics` fetches the metrics endpoint of the node (:9090). For a healthy node, this is (almost) instant. For a faulty node, we get a `ConnectionRefused` and is thus also (almost) instant.

**Comment by pierugo-dfinity** [comment]
The test is [still flaky](https://dash.zh1-idx1.dfinity.network/invocation/792ae1e4-4475-4a06-9f36-d70a6b957473?target=%2F%2Frs%2Ftests%2Fnested%3Anns_recovery_test_head_nns&targetStatus=11#1@1828), and we are actively trying to find the cause. In the meantime, let us merge this PR, which should fix the original source of flakiness.
